### PR TITLE
Reorder NCF data pipeline

### PR DESCRIPTION
**Pull request by robieta**
This PR does two things:

1) It replaces the "spillover" mechanism with a more conventional padding which is ignored by the cross entropy. @shizhiw In the last batch the padded keys all lookup at index zero. Is this hot key going to be an issue? (If so I can spread out the keys; it's just easier to read if all the paddings have the same value.)

2) It consolidates a lot of the eval generation and merges the training and eval generation. This removes some duplicate code. However, it also makes the performance properties much better:
 * Training and eval now use the same base data shards, saving a ~2GB .npz intermediate file
 * The eval set is now computed after the first training epoch, so training can begin sooner.
 * The clock start can be moved to right before the subprocess starts, because the random sampling for eval no longer happens as part of the main thread data preprocessing.

### CODE REVIEW

**Comment by robieta** [comment]
P.S. This is a prerequsite for adding logging. (Since it dictates where the tags can go.)

File: official/recommendation/ncf_main.py [hunk 1]
```diff
@@ -156,8 +156,11 @@ def run_ncf(_):
         cache_id=FLAGS.cache_id)
     num_users = ncf_dataset.num_users
     num_items = ncf_dataset.num_items
-    approx_train_steps = int(ncf_dataset.num_train_positives
-                             * (1 + FLAGS.num_neg) // FLAGS.batch_size)
+    num_train_steps = int(np.ceil(
+      FLAGS.epochs_between_evals * ncf_dataset.num_train_positives *
```
Comment (shizhiw) [inline official/recommendation/ncf_main.py]: do you need +1 here?

Comment (shizhiw) [reply official/recommendation/ncf_main.py]: nvm, you're right.

File: official/recommendation/data_async_generation.py [hunk 2]
```diff
@@ -157,75 +176,80 @@ def init_worker():
   signal.signal(signal.SIGINT, sigint_handler)
 
 
-def _construct_training_records(
+def _construct_records(
```
Comment (reedwm) [inline official/recommendation/data_async_generation.py]: Won't padding the final training batch every epoch slow down training? Why remove spillover?

File: official/recommendation/data_async_generation.py [hunk 3]
```diff
@@ -387,23 +380,32 @@ def remove_alive_file():
...
-                         eval_batch_size=eval_batch_size)
+  _construct_records(
+    is_training=True, train_cycle=train_cycle, num_neg=num_neg,
+    num_positives=num_train_positives, epochs_per_cycle=epochs_per_cycle,
+    batch_size=train_batch_size, **shared_kwargs)
+
+  # Construct evaluation set.
+  shared_kwargs["num_workers"] = num_workers
+  _construct_records(
+    is_training=False, train_cycle=-1, num_neg=rconst.NUM_EVAL_NEGATIVES,
```
Comment (reedwm) [inline official/recommendation/data_async_generation.py]: Set train_cycle to None instead

File: official/recommendation/data_preprocessing.py [hunk 4]
```diff
@@ -686,90 +652,39 @@ def input_fn(params):
...
 
 
-def make_train_synthetic_input_fn():
+def make_synthetic_input_fn(is_training):
   """Construct training input_fn that uses synthetic data."""
   def input_fn(params):
     """Generated input_fn for the given epoch."""
-    batch_size = params["batch_size"]
+    batch_size = (params["batch_size"] if is_training else
+                  params["eval_batch_size"])
```
Comment (reedwm) [inline official/recommendation/data_preprocessing.py]: The eval batch size should default to `params["batch_size"]` if not set

File: official/recommendation/data_async_generation.py [hunk 5]
```diff
@@ -104,16 +112,27 @@ def _process_shard(args):
...
 
-    negatives = stat_utils.sample_with_exclusion(
-        num_items, positive_set, n_pos * num_neg)
+    if is_training:
+      n_pos = len(positive_set)
+      negatives = stat_utils.sample_with_exclusion(
+        num_items, positive_set, n_pos * num_neg, replacement=True)
+
+    else:
+      positive_set.add(test_positive_dict[current_user])
```
[the remark this answers is not in the document]
Comment (robieta) [reply official/recommendation/data_async_generation.py]: I'll key this off of `match_mlperf`. Good catch.

File: official/recommendation/data_async_generation.py (hunk 2 above)
Comment (robieta) [reply official/recommendation/data_async_generation.py]: Rules. Given that this thing trains at ~100 batches/sec, it is unlikely to cause too much harm.

File: official/recommendation/data_async_generation.py [hunk 6]
```diff
@@ -92,8 +94,14 @@ def _process_shard(args):
...
     shard = pickle.load(f)
 
-  users = shard[movielens.USER_COLUMN]
-  items = shard[movielens.ITEM_COLUMN]
+  users = shard[rconst.TRAIN_KEY][movielens.USER_COLUMN]
+  items = shard[rconst.TRAIN_KEY][movielens.ITEM_COLUMN]
+
+  if not is_training:
+    # For eval, there is one positive which was held out from the training set.
+    test_positive_dict = {k: v for k, v in zip(
```
[the remark this answers is not in the document]
Comment (robieta) [reply official/recommendation/data_async_generation.py]: Turns out that creates a set of tuples, not a dict.

[the remark this answers is not in the document]
Comment (reedwm) [reply official/recommendation/data_async_generation.py]: I never lead you astray. You need to literal use the word `dict` instead of using the `{...}` syntax.

[the remark this answers is not in the document]
Comment (robieta) [reply official/recommendation/data_async_generation.py]: I have remedied the grievous slander to your name.

File: official/recommendation/ncf_main.py [hunk 7]
```diff
@@ -206,29 +209,40 @@ def run_ncf(_):
...
-  pred_input_fn = data_preprocessing.make_pred_input_fn(ncf_dataset=ncf_dataset)
 
+  pred_input_fn = None
   total_training_cycle = FLAGS.train_epochs // FLAGS.epochs_between_evals
   for cycle_index in range(total_training_cycle):
     tf.logging.info("Starting a training cycle: {}/{}".format(
         cycle_index + 1, total_training_cycle))
 
     # Train the model
     train_input_fn, train_record_dir, batch_count = \
```
[the remark this answers is not in the document]
Comment (robieta) [reply official/recommendation/ncf_main.py]: My inclination is that that is a non-trivial change. Lets get this in, follow up, and we can address in a subsequent PR if it turns out that that's how we want to go.
